doc/developers/contributing.rst

```diff
@@ -734,7 +736,10 @@ In general have the following in mind:
     5. Specify ``dataframe`` when "frame-like" features are being used, such
        as the column names.
     6. When specifying the data type of a list, use ``of`` as a delimiter:
-       ``list of int``.
+       ``list of int``. When the parameter supports arrays giving details about
+       the shape and/or data type and a list of such arrays, you should use
+       ``list thereof`` to be concise; e.g.
+       ``array-like of shape (n_samples,) or list thereof``.
```

Comment:
I think we should make a whole new entry for this, and also not force the use of `thereof` as there are good alternatives, and @adrinjalali and @thomasjpfan have expressed concerns. May I suggest:

> 7. When a parameter accepts a given datatype or a container of this datatype, one may use the following alternatives:
- `array-like of shape (n_samples,) or list thereof`
- `array-like of shape (n_samples,) or list of these`
- `array-like of shape (n_samples,) or list of such arrays`